packages/server-wallet/src/channel-manager.ts

```diff
@@ -0,0 +1,92 @@
+import {CreateChannelParams} from '@statechannels/client-api-schema';
+import {backOff, IBackOffOptions} from 'exponential-backoff';
+
+import {MessageServiceInterface} from './message-service/types';
+import {getMessages} from './message-service/utils';
+import {WalletObjective} from './models/objective';
+import {Wallet} from './wallet';
+export const delay = async (ms = 10): Promise<void> =>
+  new Promise(resolve => setTimeout(resolve, ms));
+
+export class ChannelManager {
+  /**
+   * Constructs a channel manager that will ensure objectives get accomplished by resending messages if needed.
+   * @param wallet The wallet to use.
+   * @param messageService  The message service to use.
+   * @param backOffIntervals At what intervals ensureObjective should retry sending messages to complete an objective.
+   * @returns A channel manager.
+   */
+  public static async create(
+    wallet: Wallet,
+    messageService: MessageServiceInterface,
+    backOffOptions?: Partial<Exclude<IBackOffOptions, 'retry'>>
+  ): Promise<ChannelManager> {
+    return new ChannelManager(wallet, messageService, backOffOptions ?? {});
+  }
+  private constructor(
+    private _wallet: Wallet,
+    private _messageService: MessageServiceInterface,
+    private _backOffOptions: Partial<IBackOffOptions>
+  ) {}
+
+  /**
+   * TODO: This is a basic implementation of createChannels.
+   * This will be cleaned up in https://github.com/statechannels/statechannels/issues/3365
+   * @param args
+   * @param numberOfChannels
+   * @returns
+   */
+  public async createChannels(args: CreateChannelParams, numberOfChannels: number): Promise<void> {
+    const createResult = await this._wallet.createChannels(args, numberOfChannels);
+
+    await this._messageService.send(getMessages(createResult));
+    await this.ensureObjectives(createResult.newObjectives);
+  }
+
+  /**
+   * Ensures that the provided objectives get completed.
+   * Will resend messages as required to ensure that the objectives get completed.
+   * @param objectives The list of objectives to ensure get completed.
+   * @returns A promise that resolves when all the objectives are completed
+   */
+  private async ensureObjectives(objectives: WalletObjective[]): Promise<void> {
+    const remaining = new Map(objectives.map(o => [o.objectiveId, o]));
+
+    const onObjectiveSucceeded = (o: WalletObjective) => {
+      if (o.objectiveId === o.objectiveId) {
+        remaining.delete(o.objectiveId);
+      }
+
+      if (remaining.size === 0) {
+        this._wallet.removeListener('objectiveSucceeded', onObjectiveSucceeded);
+      }
+    };
+
+    this._wallet.on('objectiveSucceeded', onObjectiveSucceeded);
```

Comment:
In many wallet use cases, this will exceed the default max listener count of `10`. How much should we worry about that?

Reply:
My understanding is that the MaxListener count is just there to [flag possible memory leaks ](https://nodejs.org/api/events.html#events_emitter_setmaxlisteners_n) and is not an actual limit.

It's probably prudent to set the `MaxListenersCount` to a higher number. Once that is high enough that we don't normally see it but could be triggered if we have a memory leak.

Reply:
If we set `MaxListenersCount` to a higher number, then we lose the ability to detect possible memory leaks. Should we be worried about that?

There is an alternative approach to implement `ensureObjectives`:
- currently, each `ensureObjectives` call creates an `wallet.on('objectiveSucceeded', listener)`
- instead, the wallet could manage a global set of "in progress" objectives, and set up one listener which does, roughly, `this.on('objectiveSucceeded', objective => this.removeInProgressObjective(objective.id))`

Is that, or some variation of that, preferable?

Reply:
> If we set `MaxListenersCount` to a higher number, then we lose the ability to detect possible memory leaks. Should we be worried about that?

Setting `MaxListenerCount` to a higher value just means it would take longer to detect the memory leak. If we're adding listeners and never removing them then eventually we'll hit `MaxListenerCount` even if it's a bit higher.

> There is an alternative approach to implement `ensureObjectives`:
> 
> * currently, each `ensureObjectives` call creates an `wallet.on('objectiveSucceeded', listener)`
> * instead, the wallet could manage a global set of "in progress" objectives, and set up one listener which does, roughly, `this.on('objectiveSucceeded', objective => this.removeInProgressObjective(objective.id))`. then, `ensureObjectives` would check `wallet.anyInProgress?(objectivesOfInterest)`
> 
> Is that, or some variation of that, preferable?

Instead of calling a potential  `wallet.anyInProgress?(objectivesOfInterest)` couldn't we just query the objectives? That way we don't need the listeners at all.

Reply:
> Setting MaxListenerCount to a higher value just means it would take longer to detect the memory leak. 

That's still a drawback! If it takes 12 hours under some load, to detect the memory leak, but we never leave a wallet running for 12 hours, we won't detect it.

> couldn't we just query the objectives?

Can you explain your suggestion in words or pseudocode? I think what you're suggesting introduces unnecessary DB calls, which can be a problem under load.

Reply:
> > Setting MaxListenerCount to a higher value just means it would take longer to detect the memory leak.
> 
> That's still a drawback! If it takes 12 hours under some load, to detect the memory leak, but we never leave a wallet running for 12 hours, we won't detect it.

It seems like a slightly convoluted constraint that we require the wallet have at most 10 listeners just for the purpose of a warning message?

I guess I feel that the benefits of the `maxListener` check are pretty slim. It seems like a really rough heuristic for a memory leak. 

Also by moving more logic out of listeners doesn't that mean we lessen our ability to detect a memory leak? IE: If the wallet fails to properly remove an objective and keeps eating up memory with it's collection of `InProgress` objectives there'd be nothing to catch that. 

> > couldn't we just query the objectives?
> 
> Can you explain your suggestion in words or pseudocode? I think what you're suggesting introduces unnecessary DB calls, which can be a problem under load.

This is what I sketched out but it probably requires too many DB calls:
```
  private async ensureObjectives(objectiveIds: string[]): Promise<void> {
    try {
      await backOff(
        async () => {
          const {outbox} = await this._wallet.syncObjectives(objectiveIds);
          await this._messageService.send(getMessages(outbox));

          if (await this.objectivesIncomplete(objectiveIds)) {
            // Throwing an error indicates to the backoff library that the task is not complete
            throw new Error('Still objectives to complete');
          }
        },

        {...this._backOffOptions, retry: async () => this.objectivesIncomplete(objectiveIds)}
      );
    } catch (error) {
      const remaining = await this.getIncompleteObjectives(objectiveIds);
      this._wallet.logger.error('Unable to ensure objectives', {
        remaining: Array.from(remaining.keys()),
      });
      throw new Error('Unable to ensure objectives');
    }
  }
```

Reply:
> > > Setting MaxListenerCount to a higher value just means it would take longer to detect the memory leak.
> > 
> > 
> > That's still a drawback! If it takes 12 hours under some load, to detect the memory leak, but we never leave a wallet running for 12 hours, we won't detect it.
> 
> It seems like a slightly convoluted constraint that we require the wallet have at most 10 listeners just for the purpose of a warning message?
> 
> I guess I feel that the benefits of the `maxListener` check are pretty slim. It seems like a really rough heuristic for a memory leak.

It is really easy to forget to do things like [this](https://github.com/statechannels/statechannels/commit/ee2cd39f84c672aed65303956e241d06b4941b3b#diff-d097197de4d5617abf526bc232c8ccd50cbebc3e695669e910b5cc77183ba6c3R67). The warning would point out that we haven't cleaned up in some code paths. To trigger the warning, we'd need an environment that hits all code paths "enough times".

To me, this is an indication that we should aim to _avoid using event emitters & listeners for managing internal logic_. I think it enables a lot of footguns, and makes things hard to debug. Since there's a general feeling that `wallet.ts` should not be an event emitter, I think it's important to consider what it would take to make that happen. 

> Also by moving more logic out of listeners doesn't that mean we lessen our ability to detect a memory leak? IE: If the wallet fails to properly remove an objective and keeps eating up memory with it's collection of `InProgress` objectives there'd be nothing to catch that.

A test can catch that, by asserting on the contents of `inProgress` objectives!

> This is what I sketched out but it probably requires too many DB calls:
> 
> (SNIP)
> 
> Both `this.objectivesIncomplete(objectiveIds)` and `this.getIncompleteObjectives(objectiveIds);` would be performing a query.

Yes, I think this is undesirable.

--- 

I think it's worth looking at this work from a higher level: we are _implicitly_ committed to a single wallet instance being responsible for managing channels. Why? Because two such wallet instances would both use the same private key to make blockchain transactions, and we'd result in nonce issues. We'd also have two chain services listening to, and trying to respond to, the same event.

Is it worth making this implicit assumption explicit? The benefit to doing that is, we can then manage some things in memory, such as a list of "in progress objectives". Keeping such a list in memory may allow us to make `wallet.ts` _stop being an event emitter_.